### ISSUE TITLE
Add storage from Android content provider

### ISSUE THREAD

**Issue by Senliast** [TeamAmaze/AmazeFileManager#4019]
**Is your feature request related to a problem? Please describe.**
Currently not possible to add storage, that is provided by another app (for example, SMB file explorer) that will be shown in default Android file manager / content provider.

**Describe the solution you'd like**
Add an option to add storage from Android content provider (default file manager). I.e. "use this folder" from Android content provider.

**Describe alternatives you've considered**
nv

**Additional context**
nv

**Comment by VishnuSanal** [comment]
@Senliast thanks for reaching out. but, I didn't quite get that, did you mean accessing external storage like OTG devices? or SMB devices? we support both! please clarify your feature request, thanks.

**Comment by Senliast** [comment]
@VishnuSanal hi, i will show an example. I have an app installed "CIFS Documents Provider" - it connects to an SMB share and then provides it in the Android content provider. I.e. it's not a file manager by itself, it only connects to the share and then makes it available to other apps. It can then be selected in another app, that supports it. For example, it Total Commander, i can add it as a location. Here is screenshot:
![Screenshot_20231220-175712_Dateien](https://github.com/TeamAmaze/AmazeFileManager/assets/23285435/82afe844-4024-4810-9cef-fa44ffdd792e)

**Comment by VishnuSanal** [comment]
I don't think I completely follow you, but I think what you are referring to is to be able to list FTP/SMB/SFTP locations on other applications as target locations. I feel like this would be redundant in Amaze because we are a file manager in itself! but, this is a nice-to-have is open to contributions. also, any idea on how to implement this feature is invited too.

**Comment by e-t-l** [comment]
I'm not sure if it's the same as FTP/SMB actually, since afaik those are network-based sharing protocols. I believe what is being discussed here is an Android-specific protocol called _[content providers (class DocumentsProvider)](https://developer.android.com/reference/android/provider/DocumentsProvider)_. This is used by apps such as [Shelter](https://github.com/achalmgucker/Shelter) to provide access to the Work Profile filesystem, and [RoundSync](https://github.com/newhinton/Round-Sync) to provide access to cloud-based files as if they were local. The apps for Google Drive, OneDrive, DropBox, etc also do this. 

The only *FOSS* app I have seen that includes access to content providers is actually the default AOSP "Files" all (aka [DocumentsUI](https://android.googlesource.com/platform/packages/apps/DocumentsUI/+/refs/heads/main), which as it says, is not directly accessible after Android 12. However, the [GrapheneOS Files app](https://github.com/GrapheneOS/platform_packages_apps_DocumentsUI) is built on DocumentsUI and *is* user accessible, so that might prove a useful reference.
AOSP also has a developer guide for how to access content providers that might help: https://developer.android.com/guide/topics/providers/document-provider

Like I said, no other FOSS app (that aren't the very basic Android Files app) does this, so if Amaze could access content providers, that would give it a huge advantage.

**Comment by slimhk45** [comment]
+1 for implementing `DocumentsProvider` like the DocumentUI-based GrapheneOS Files.

I need to copy folders from cloud storage to local storage to initialize AutoSync. That would be super useful!

**Comment by Senliast** [comment]
There is for example an app EDS or EDS NG, that allows you to access VeraCrypt-volumes on Android. And after it mounts the volume, it will be available in this Documents Provider. So, its usage area is more wider than SMB/OTG, it basically allows to interact with any apps that share their private content. So it would be very usefull.

**Comment by slimhk45** [comment]
> There is for example an app EDS or EDS NG, that allows you to access VeraCrypt-volumes on Android. And after it mounts the volume, it will be available in this Documents Provider. So, its usage area is more wider than SMB/OTG, it basically allows to interact with any apps that share their private content. So it would be very usefull.

Yes, basically all cloud storage providers (Google Drive, Proton Drive etc.) and virtual disk encryption tool (DroidFS, Cryptomator) expose their storage using `DocumentsProvider` from [Storage Access Framework](https://developer.android.com/guide/topics/providers/create-document-provider). 

PS: use [DroidFS](https://github.com/hardcore-sushi/DroidFS) (free) or Cryptomator (paid) instead of EDS NG. They have better UX and a faster file-based encryption.

**Comment by e-t-l** [comment]
Following up on my previous comment, I believe DocumentsProvider storage is accessible through basic SAF, so it may be simplest to make that part of Amaze be basically just a wrapper for SAF.
